libs/basic_execution/include/hpx/basic_execution/receiver.hpp

```diff
@@ -0,0 +1,365 @@
+    template <typename R, typename E>
+    void set_error(R&& r, E&& e);
+#endif
+
+    /// Receiving values from asynchronous computations is handled by the `receiver`
+    /// concept. A `receiver` needs to be able to receive an error or be marked as
+    /// being cancelled. As such, the Receiver concept is defined by having the
+    /// following two customization points defined, which form the completion-signal
+    /// operations:
+    ///     * `hpx::basic_execution::set_done`
+    ///     * `hpx::basic_execution::set_error`
+    ///
+    /// Those two functions denote the completion-signal operations. The receiver
+    /// contract is as follows:
+    ///     * None of a receiver's completion-signal operation shall be invoked
+    ///       before `hpx::basic_execution::start` has been called on the operation
+    ///       state object that was returned by connecting a receiver to a sender
+    ///       `hpx::basic_execution::connect`.
+    ///     * Once `hpx::basic_execution::start` has been called on the operation
+    ///       state object, exactly one of the receiver's completion-signal operation
+    ///       shall complete without an exception before the receiver is destroyed
+    ///
+    /// Once one of the receiver's completion-signal operation has been completed
+    /// without throwing an exception, the receiver contract has been satisfied.
+    /// In other words: The asynchronous operation has been completed.
+    ///
+    /// \see hpx::basic_execution::is_receiver_of
+    template <typename T, typename E = std::exception_ptr>
+    struct is_receiver;
+
+    /// The `receiver_of` concept is a refinement of the `receiver` concept by
+    /// requiring one additional completion-signal operation:
+    ///     * `hpx::basic_execution::set_value`
+    ///
+    /// This completion-signal operation adds the following to the receiver's
+    /// contract:
+    ///     * If `hpx::basic_execution::set_value` exits with an exception, it
+    ///       is still valid to call `hpx::basic_execution::set_error` or
+    ///       `hpx::basic_execution::set_done`
+    ///
+    /// \see hpx::basic_execution::is_receiver
+    template <typename T, typename... As>
+    struct is_receiver_of;
+
+    HPX_INLINE_CONSTEXPR_VARIABLE struct set_value_t
+    {
+#define HPX_BASIC_EXECUTION_RECEIVER_SET_VALUE_EXPRESSION                      \
+    hpx::functional::tag_invoke(                                               \
+        *this, std::forward<Receiver>(rcv), std::forward<Values>(values)...)
+        template <typename Receiver, typename... Values>
+        constexpr HPX_FORCEINLINE auto tag_invoke_impl(
+            std::integral_constant<bool, true> /* is tag invocable */,
+            Receiver&& rcv, Values&&... values) const
+            noexcept(
+                noexcept(HPX_BASIC_EXECUTION_RECEIVER_SET_VALUE_EXPRESSION))
+                -> decltype(HPX_BASIC_EXECUTION_RECEIVER_SET_VALUE_EXPRESSION)
+        {
```

Comment:
Wouldn't it be (slightly) simpler to use `decltype(auto)` as the return type instead of spelling out the fulle `-> decltype(…)`?

Reply:
For some of the customization points (the ones without the `noexcept()` clause) this would also remove the need for defining the macro.

Reply:
unfortunately, that's not going to work... it needs to be spelled out fully for SFINAE to work so that the dispatching works properly.
Unless we require C++17 and up, IIRC.